File: frontend-next/lib/record/recording-form-metadata.ts
import type { RecordingFormDefinition } from '@/lib/record/recording-form-registry'
import { lifecycleForForm } from '@/lib/record/recording-form-lifecycle'
import { sccifAlignmentForForm } from '@/lib/record/recording-form-sccif-alignment'
import {
  ACTIONS_FOLLOW_UP_PROMPT,
  ADULT_RESPONSE_SECTION_PROMPT,
  CHILD_VOICE_SECTION_PROMPT,
  PLAN_IMPACT_CHECK_PROMPT,
  THERAPEUTIC_LANGUAGE_GUIDANCE
} from '@/lib/record/recording-form-therapeutic-defaults'

export type FormalRouteClassification =
  | 'SUPPORTED_NOW'
  | 'REVIEW_THEN_SUPPORTED'
  | 'DRAFT_ONLY'
  | 'ROUTE_HINT_ONLY'
  | 'NEEDS_FORMAL_BACKEND'

export type RecordingFormRecordMetadata = {
  record_date: string
  event_date: string
  event_time?: string
  written_by_user_id?: string
  written_by_name?: string
  written_by_role?: string
  last_edited_by_user_id?: string
  last_edited_by_name?: string
  reviewed_by_user_id?: string
  reviewed_by_name?: string
  reviewed_at?: string
  signed_off_by_user_id?: string
  signed_off_by_name?: string
  signed_off_at?: string
  child_id?: number
  home_id?: number
  form_id: string
  form_type: string
  category?: string
  source: 'recording_workspace' | 'manager_review' | 'formal_submit'
  status: 'draft' | 'ready_for_review' | 'submitted' | 'signed_off' | 'archived' | 'read_only'
  review_status?: string
  manager_review_required: boolean
  safeguarding_review_required: boolean
  privacy_sensitive: boolean
  child_voice_present?: boolean
  adult_response_present?: boolean
  actions_required?: boolean
  chronology_event_id?: string
  archive_record_id?: string
  plan_impact_ids?: string[]
  lifeecho_suggestion_ids?: string[]
  formal_route_classification: FormalRouteClassification
  formal_route_hint?: string
  formal_route_warning?: string
  is_signed_off: boolean
  is_editable: boolean
  editability_note?: string
  therapeutic_flags: {
    child_voice_prompt: string
    adult_response_prompt: string
    actions_follow_up_prompt: string
    plan_impact_check_prompt: string
    language_guidance: string[]
  }
  lifecycle: ReturnType<typeof lifecycleForForm>
  sccif_alignment: ReturnType<typeof sccifAlignmentForForm>
}

const SUPPORTED_NOW_IDS = new Set([
  'daily-note',
  'incident',
  'keywork',
  'keywork-direct-work',
  'family-time',
  'education-note',
  'health-appointment',
  'missing-episode',
  'missing'
])

const REVIEW_THEN_IDS = new Set([
  'safeguarding-concern',
  'physical-intervention',
  'injury-body-map',
  'body-map',
  'medication-error',
  'medication-note-error',
  'return-conversation',
  'disclosure',
  'allegation',
  'police-involvement',
  'hospital-emergency',
  'child-on-child-concern',
  'exploitation-concern',
  'room-search',
  'complaint-concern'
])

const ROUTE_HINT_IDS = new Set([
  'handover',
  'child-voice',
  'health-medication',
  'care-plan-update',
  'staff-supervision',
  'action-plan-note',
  'evidence-document',
  'reg44-evidence',
  'reg45-evidence',
  'wishes-and-feelings',
  'medication-administration'
])

const NEEDS_BACKEND_IDS = new Set([
  'safeguarding-concern',
  'medication-error',
  'physical-intervention',
  'body-map',
  'injury-body-map',
  'missing-episode',
  'complaint-concern',
  'room-search',
  'behaviour-support'
])

export function formalRouteClassificationForForm(form: RecordingFormDefinition): FormalRouteClassification {
  if (SUPPORTED_NOW_IDS.has(form.id) || form.workflowStatus === 'formal_submit_supported') {
    return 'SUPPORTED_NOW'
  }
  if (REVIEW_THEN_IDS.has(form.id) || form.workflowStatus === 'safeguarding_sensitive') {
    return 'REVIEW_THEN_SUPPORTED'
  }
  if (ROUTE_HINT_IDS.has(form.id) || form.workflowStatus === 'opens_existing_workflow') {
    return 'ROUTE_HINT_ONLY'
  }
  if (NEEDS_BACKEND_IDS.has(form.id)) {
    return 'NEEDS_FORMAL_BACKEND'
  }
  return 'DRAFT_ONLY'
}

export function formalRouteHintForClassification(
  classification: FormalRouteClassification,
  formTitle: string
): { hint: string; warning?: string } {
  switch (classification) {
    case 'SUPPORTED_NOW':
      return {
        hint: `Formal submit supported for ${formTitle} when a child is selected and review rules are met.`
      }
    case 'REVIEW_THEN_SUPPORTED':
      return {
        hint: `Manager or safeguarding review required before ${formTitle} can become a formal record.`,
        warning: 'Do not treat as completed until review is recorded.'
      }
    case 'ROUTE_HINT_ONLY':
      return {
        hint: `Prepare ${formTitle} here, then complete in the existing formal workflow route.`,
        warning: 'Automatic formal record creation is not wired from this workspace.'
      }
    case 'NEEDS_FORMAL_BACKEND':
      return {
        hint: `${formTitle} — dedicated formal backend route not fully wired from workspace.`,
        warning: 'Save as draft and use the correct module, or manager review queue when ready.'
      }
    case 'DRAFT_ONLY':
    default:
      return {
        hint: `${formTitle} can be drafted here. Formal workflow not wired yet — save draft and use correct route.`,
        warning: 'No automatic formal record will be created on submit.'
      }
  }
}

export function todayIsoDate(): string {
  return new Date().toISOString().slice(0, 10)
}

export function buildDefaultFormRecordMetadata(
  form: RecordingFormDefinition,
  options?: {
    childId?: number
    homeId?: number
    writtenByUserId?: string
    writtenByName?: string
    writtenByRole?: string
    eventDate?: string
    recordDate?: string
    isSignedOff?: boolean
  }
): RecordingFormRecordMetadata {
  const classification = formalRouteClassificationForForm(form)
  const routeCopy = formalRouteHintForClassification(classification, form.title)
  const isSignedOff = options?.isSignedOff ?? false
  return {
    record_date: options?.recordDate || todayIsoDate(),
    event_date: options?.eventDate || todayIsoDate(),
    written_by_user_id: options?.writtenByUserId,
    written_by_name: options?.writtenByName,
    written_by_role: options?.writtenByRole,
    child_id: options?.childId,
    home_id: options?.homeId,
    form_id: form.id,
    form_type: form.workspaceType || form.id,
    category: form.category,
    source: 'recording_workspace',
    status: isSignedOff ? 'read_only' : 'draft',
    manager_review_required: form.requiresManagerReview,
    safeguarding_review_required:
      form.safeguardingSensitive || form.workflowStatus === 'safeguarding_sensitive',
    privacy_sensitive: form.privacySensitive,
    formal_route_classification: classification,
    formal_route_hint: routeCopy.hint,
    formal_route_warning: routeCopy.warning,
    is_signed_off: isSignedOff,
    is_editable: !isSignedOff,
    editability_note: isSignedOff
      ? 'This record is signed off and cannot be edited directly. Create an addendum via the correction workflow when available.'
      : undefined,
    therapeutic_flags: {
      child_voice_prompt: CHILD_VOICE_SECTION_PROMPT,
      adult_response_prompt: ADULT_RESPONSE_SECTION_PROMPT,
      actions_follow_up_prompt: ACTIONS_FOLLOW_UP_PROMPT,
      plan_impact_check_prompt: PLAN_IMPACT_CHECK_PROMPT,
      language_guidance: [...THERAPEUTIC_LANGUAGE_GUIDANCE]
    },
    lifecycle: lifecycleForForm(form.id, form.category),
    sccif_alignment: sccifAlignmentForForm(
      form.id,
      form.category,
      form.relatedQualityStandards,
      form.relatedEvidenceAreas
    )
  }
}

export function parseFormRecordMetadata(
  metadata?: Record<string, unknown> | null
): RecordingFormRecordMetadata | null {
  if (!metadata) return null
  const raw = metadata.form_record as RecordingFormRecordMetadata | undefined
  if (!raw || typeof raw !== 'object') return null
  return raw
}

export function mergeFormRecordMetadataPatch(
  existing: Record<string, unknown> | undefined,
  patch: Partial<RecordingFormRecordMetadata>
): Record<string, unknown> {
  const current = parseFormRecordMetadata(existing) || ({} as RecordingFormRecordMetadata)
  return {
    ...existing,
    form_record: { ...current, ...patch }
  }
}

File: schemas/recording_form_metadata.py
"""Universal care-record metadata for recording workspace drafts and signed-off records."""

from __future__ import annotations

from datetime import date, datetime
from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field

RecordingFormMetadataSource = Literal["recording_workspace", "manager_review", "formal_submit", "import"]
RecordingFormMetadataStatus = Literal[
    "draft",
    "ready_for_review",
    "submitted",
    "signed_off",
    "archived",
    "read_only",
]

ArchiveBehaviour = Literal["signed_off_only", "restricted_summary", "draft_never", "not_applicable"]
ChronologyBehaviour = Literal["create_story_event", "link_only", "restricted_summary", "not_applicable"]
PlanImpactBehaviour = Literal[
    "health_plan",
    "education_plan",
    "risk_assessment",
    "family_time_plan",
    "behaviour_support_plan",
    "medication_plan",
    "missing_plan",
    "safeguarding_plan",
    "none",
]
LifeEchoBehaviour = Literal["positive_safe_only", "never_auto", "review_required", "not_applicable"]
FormalRouteClassification = Literal[
    "SUPPORTED_NOW",
    "REVIEW_THEN_SUPPORTED",
    "DRAFT_ONLY",
    "ROUTE_HINT_ONLY",
    "NEEDS_FORMAL_BACKEND",
]

SCCIF_EVIDENCE_DISCLAIMER = (
    "May support evidence for inspection themes — not a compliance or grading statement."
)


class RecordingFormLifecycleBehaviour(BaseModel):
    model_config = ConfigDict(extra="ignore")

    archive_behaviour: ArchiveBehaviour = "signed_off_only"
    chronology_behaviour: ChronologyBehaviour = "create_story_event"
    plan_impact_behaviour: PlanImpactBehaviour = "none"
    lifeecho_behaviour: LifeEchoBehaviour = "review_required"


class RecordingFormSccifAlignment(BaseModel):
    model_config = ConfigDict(extra="ignore")

    quality_standards: list[str] = Field(default_factory=list)
    sccif_evidence_areas: list[str] = Field(default_factory=list)
    regulatory_relevance: list[str] = Field(default_factory=list)
    inspection_evidence_type: str | None = None
    alignment_note: str = SCCIF_EVIDENCE_DISCLAIMER


class RecordingFormTherapeuticFlags(BaseModel):
    model_config = ConfigDict(extra="ignore")

    child_voice_prompt: str = "What was the child communicating?"
    adult_response_prompt: str = "What did adults notice and do to help?"
    actions_follow_up_prompt: str = "What follow-up or actions are needed?"
    plan_impact_check_prompt: str = "Does this affect a plan or risk assessment?"
    review_check_prompt: str = "Does this need manager or safeguarding review?"
    language_guidance: list[str] = Field(default_factory=list)


class RecordingFormMetadata(BaseModel):
    """Canonical metadata block stored in draft.metadata.form_record or top-level fields."""

    model_config = ConfigDict(extra="ignore")

    record_date: date | None = None
    event_date: date | None = None
    event_time: str | None = None
    written_by_user_id: str | None = None
    written_by_name: str | None = None
    written_by_role: str | None = None
    created_at: datetime | str | None = None
    updated_at: datetime | str | None = None
    last_edited_by_user_id: str | None = None
    last_edited_by_name: str | None = None
    reviewed_by_user_id: str | None = None
    reviewed_by_name: str | None = None
    reviewed_at: datetime | str | None = None
    signed_off_by_user_id: str | None = None
    signed_off_by_name: str | None = None
    signed_off_at: datetime | str | None = None
    child_id: int | None = None
    home_id: int | None = None
    form_id: str | None = None
    form_type: str | None = None
    category: str | None = None
    source: RecordingFormMetadataSource = "recording_workspace"
    status: RecordingFormMetadataStatus = "draft"
    review_status: str | None = None
    manager_review_required: bool = False
    safeguarding_review_required: bool = False
    privacy_sensitive: bool = False
    therapeutic_flags: RecordingFormTherapeuticFlags = Field(default_factory=RecordingFormTherapeuticFlags)
    child_voice_present: bool | None = None
    adult_response_present: bool | None = None
    actions_required: bool | None = None
    chronology_event_id: str | None = None
    archive_record_id: str | None = None
    plan_impact_ids: list[str] = Field(default_factory=list)
    lifeecho_suggestion_ids: list[str] = Field(default_factory=list)
    lifecycle: RecordingFormLifecycleBehaviour = Field(default_factory=RecordingFormLifecycleBehaviour)
    sccif_alignment: RecordingFormSccifAlignment = Field(default_factory=RecordingFormSccifAlignment)
    formal_route_classification: FormalRouteClassification = "DRAFT_ONLY"
    formal_route_hint: str | None = None
    formal_route_warning: str | None = None
    structured_template_id: str | None = None
    is_signed_off: bool = False
    is_editable: bool = True
    editability_note: str | None = None


def merge_form_metadata(
    existing: dict[str, Any] | None,
    patch: dict[str, Any],
) -> dict[str, Any]:
    """Shallow-merge form_record metadata preserving nested therapeutic_flags."""
    base = dict(existing or {})
    form_record = dict(base.get("form_record") or {})
    for key, value in patch.items():
        if key == "therapeutic_flags" and isinstance(value, dict):
            merged_flags = dict(form_record.get("therapeutic_flags") or {})
            merged_flags.update(value)
            form_record["therapeutic_flags"] = merged_flags
        elif key == "lifecycle" and isinstance(value, dict):
            merged_lifecycle = dict(form_record.get("lifecycle") or {})
            merged_lifecycle.update(value)
            form_record["lifecycle"] = merged_lifecycle
        elif key == "sccif_alignment" and isinstance(value, dict):
            merged_sccif = dict(form_record.get("sccif_alignment") or {})
            merged_sccif.update(value)
            form_record["sccif_alignment"] = merged_sccif
        else:
            form_record[key] = value
    base["form_record"] = form_record
    return base


def default_metadata_for_form(
    *,
    form_id: str,
    form_type: str,
    category: str | None,
    child_id: int | None = None,
    home_id: int | None = None,
    written_by_user_id: str | None = None,
    written_by_name: str | None = None,
    written_by_role: str | None = None,
    manager_review_required: bool = False,
    safeguarding_review_required: bool = False,
    privacy_sensitive: bool = False,
    lifecycle: RecordingFormLifecycleBehaviour | None = None,
    sccif_alignment: RecordingFormSccifAlignment | None = None,
    formal_route_classification: FormalRouteClassification = "DRAFT_ONLY",
) -> dict[str, Any]:
    """Build metadata dict suitable for draft.metadata."""
    today = date.today().isoformat()
    meta = RecordingFormMetadata(
        record_date=date.today(),
        event_date=date.today(),
        written_by_user_id=written_by_user_id,
        written_by_name=written_by_name,
        written_by_role=written_by_role,
        child_id=child_id,
        home_id=home_id,
        form_id=form_id,
        form_type=form_type,
        category=category,
        manager_review_required=manager_review_required,
        safeguarding_review_required=safeguarding_review_required,
        privacy_sensitive=privacy_sensitive,
        lifecycle=lifecycle or RecordingFormLifecycleBehaviour(),
        sccif_alignment=sccif_alignment or RecordingFormSccifAlignment(),
        formal_route_classification=formal_route_classification,
        is_editable=True,
        editability_note=None,
    )
    return {"form_record": meta.model_dump(mode="json")}
